### PR TITLE
Allow to reset selected received requests in Hermes Mock

### PR DESCRIPTION
**Pull request by platan**
This PR adds option to reset selected received requests in Hermes Mock. Such option is useful in parallel execution of tests when you don't want to reset all requests (because this will affects all tests).

### CODE REVIEW

File: hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockAvroTest.groovy [hunk 1]
```diff
@@ -288,6 +288,33 @@ class HermesMockAvroTest extends Specification {
...
+        then:
+        hermes.query().countAvroMessages(topicName) == 0
+    }
+
+    def "should not reset received requests with Avro messages when predicate does not match"() {
+        given:
+        def topicName = "my-test-avro-topic"
+        hermes.define().avroTopic(topicName)
+        publish(topicName, new TestMessage("test-key", "test-value"))
+        def requestCountBeforeRest = hermes.query().countAvroMessages(topicName)
```
[the remark this answers is not in the document]
Comment (faderskd) [reply hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockAvroTest.groovy:309]: Rest -> Reset

File: hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockJsonTest.groovy [hunk 2]
```diff
@@ -106,6 +106,33 @@ class HermesMockJsonTest extends Specification {
...
+        then:
+        hermes.query().countAvroMessages(topicName) == 0
+    }
+
+    def "should not reset received requests with JSON messages when predicate does not match"() {
+        given:
+        def topicName = "my-test-json-topic"
+        hermes.define().avroTopic(topicName)
+        publishJson(topicName, new TestMessage("test-key", "test-value").asJson())
+        def requestCountBeforeRest = hermes.query().countAvroMessages(topicName)
```
Comment (faderskd) [inline hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockJsonTest.groovy:127]: same as above :)

File: hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockAvroTest.groovy [hunk 3]
```diff
@@ -288,6 +288,33 @@ class HermesMockAvroTest extends Specification {
             1 == hermes.query().countMatchingAvroMessages(topicName, schema, TestMessage, filter)
     }
 
+    def "should reset received requests with Avro messages"() {
+        given:
+        def topicName = "my-test-avro-topic"
+        hermes.define().avroTopic(topicName)
+        publish(topicName, new TestMessage("test-key", "test-value"))
+
```
Comment (faderskd) [inline hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockAvroTest.groovy:296]: I know it is already tested in other scenario but maybe we can make an assertion in the `given` section just for readability:
`hermes.query().countAvroMessages(topicName) == 1`

File: hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockAvroTest.groovy (hunk 1 above)
[the remark this answers is not in the document]
Comment (platan) [reply hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockAvroTest.groovy:309]: No, this variable is not required. I've changed assertion - it checks if count is 1.

File: hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockAvroTest.groovy (hunk 3 above)
Comment (platan) [reply hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockAvroTest.groovy:296]: Good idea. I've added this assertion.

**Comment by platan** [comment]
@faderskd Thank you for review. All issues resolved.

File: hermes-mock/src/main/java/pl/allegro/tech/hermes/mock/HermesMockQuery.java [hunk 4]
```diff
@@ -123,4 +128,14 @@ public <T> Optional<T> lastAvroMessageAs(String topicName, Schema schema, Class<
         return lastRequest(topicName)
                 .map(req -> hermesMockHelper.deserializeAvro(req, schema, clazz));
     }
+
+    public <T> void resetReceivedAvroRequests(String topicName, Schema schema, Class<T> clazz, Predicate<T> predicate) {
```
Comment (pitagoras3) [inline hermes-mock/src/main/java/pl/allegro/tech/hermes/mock/HermesMockQuery.java:132]: I'm thinking if `HermesMockQuery` class is good place for these functions. This class responsibility is to `query` Hermes mock, and return some data, not to modify internal state of Hermes mock. 

We have "similar" method `resetReceivedRequest` in `HermesMock` class. Maybe we can place these methods there? What do you think @faderskd?

Comment (faderskd) [reply hermes-mock/src/main/java/pl/allegro/tech/hermes/mock/HermesMockQuery.java:132]: good catch, totally agree :)

Comment (platan) [reply hermes-mock/src/main/java/pl/allegro/tech/hermes/mock/HermesMockQuery.java]: Sure. Done.

File: hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockTest.groovy [hunk 5]
```diff
@@ -276,11 +283,73 @@ class HermesMockTest extends Specification {
...
+        publish(topicName, new TestMessage("test-key", "test-value").asJson())
+
+        when:
+        hermes.resetReceivedJsonRequests(topicName, TestMessage, {it -> it.key == "different-test-key"})
+
+        then:
+        hermes.query().countJsonMessages(topicName, TestMessage) == 1
+    }
+
     def publish(String topic) {
```
Comment (pitagoras3) [inline hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockTest.groovy:340]: Can I ask you to change this function name to `publishJson`?

File: hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockTest.groovy [hunk 6]
```diff
@@ -276,11 +283,73 @@ class HermesMockTest extends Specification {
...
         publish(topic, TestMessage.random().asJson())
     }
 
     def publish(String topic, String body) {
         publisher.publish(topic, body)
     }
+
+    def publish(String topic, TestMessage message) {
```
Comment (pitagoras3) [inline hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockTest.groovy:348]: And this function to `publishAvro`, to make it clear what is send :)

File: hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockTest.groovy (hunk 5 above)
Comment (platan) [reply hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockTest.groovy]: Done.

File: hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockTest.groovy (hunk 6 above)
Comment (platan) [reply hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockTest.groovy]: Good idea. Done.

**Review by pitagoras3 — APPROVED** [review]

**Review by faderskd — APPROVED** [review]

**Pull request merged**